### ISSUE TITLE
[Bug] rc --follow-location fails when cursor is at word end

### ISSUE THREAD

**Issue by papadokolos** [Andersbakken/rtags#1032]
### Please mark appropriate
 - rtags (rdm/rc/rp)

### Problem description
Executing `rc --follow-location` with a cursor position at the end of a word returns 32, whereas executing the same command with any other position within the same word operates successfully without errors.

#### Expected behavior
I expect `rc --follow-location` to behave the same way, no matter if the given position is at the start, middle, or end of a word.

#### Actual behavior
Executing `rc --follow-location` (whether directly or with emacs) returns with an exit code of 32 (general error) whenever the given file position (row and column) directs to an end of a word.

In the picture below I'm trying to follow the locations of three different symbols. You can easily tell that something is not right when the position parameter directs to the end of a wordt:
![rtags_bug](https://user-images.githubusercontent.com/13286080/29581283-6eea9fd8-8781-11e7-8ef9-aabda166b740.png)

This issue occurs in all calls to `rc --follow-location`, whether the expected result is a function, a variable, or a namespace, and whether it is in the same file or in a different one.

### Environment
- **Operating system:**
  Ubuntu 14.04 32bit (VirtualBox) 

- **LLVM/Clang version:**
  3.8.0

**Comment by Andersbakken** [comment]
I actually think this is the correct behavior. 

The character after the variable could be an operator.

E.g.

std::string str;
str+="foo";

if you do -f on line 2 character 4 (+) you should jump to the += operator for string, not str

**Comment by papadokolos** [comment]
I see your point. When the cursor is between symbols, the symbol on the right should take precedence over the symbol on the left. However, what about cases when there is no symbol on the right?

Consider these exapmples, where the cursor position is represented with `|` character:
1. The character on the right of the cursor dictates with no doubt that the cursor is not at the start of a symbol:
    ```c++
    std::string str|; // There is no symbol on the right. Does it really makes sense to fail here?
    std::string str2("text")|; // Same case, but more complicated

    /* Assume such operator exists, for the examples below */
    int operator+(const std::string& lhs, int rhs);

    int result = str+|2; // Symbols can't start with a numerical character, so the user clearly wants the symbol on the left

    int num = 2;
    int result2 = str +| num; // The operator+ is delimited by a space, so it is more reasonable to act on the left symbol instead of failing
    str| += "foo"; // Same example as yours, but space delimited! (which justifies different behaviour IMO)
    ```
2. _rdm_ is acting (almost) exactly as I expect it to in these cases:
    ```c++
    char character = 'a';
    int times_of_character = 4;
    std::string str1(times_of_character, character|); // Location is followed successfully!! amazing!
    std::string str2(times_of_character, |character); // Works as usual
    std::string str3(times_of_character|, character); // Works as well!! I'm delighted
    std::string str4(|times_of_character, character); // Fails! but why?

    /* Now, the same case with one paramter */
    const char* text = "text";
    std::string str5(text|); // Location is followed successfully!! amazing!
    std::string str6(|text); // Fails! is it a bug?
    ```
    These examples prove that there is some basic support for the behaviour I'm requesting here (and also a possible bug?)

To sum up, I agree with you that when in a doubt, the symbol on the right should be preffered. In spite of that, it is pity to return an error when there is a perfectly legitimate option to try - to follow a symbol on the left. It is very reasonable in my opinion. I can't even count the number of times where my cursor reached the end of a symbol, and I had to reluctantly move back one character just to perform `rc --follow-location`.

**Comment by papadokolos** [comment]
Eager to create a pull request, I spent the whole day trying to understand the code responsible for following symbol location.

I figured that it involves `Project::findSymbol()`,  `FileMap::lowerbound()` and `FileMap::valueAt()`, but didn't manage to go any further. I even tried to debug the code but still couldn't understand the logic taking place in there.

I would love to get some explanation about this.
